File: watchtower-core/src/main/java/dev/mcstatus/watchtower/core/analyze/CrashNarrator.java
package dev.mcstatus.watchtower.core.analyze;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import dev.mcstatus.watchtower.core.collect.MixinConfigIndex;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Plain-English crash narratives from parsed crash report fields.
 */
public final class CrashNarrator {

    public record Narrative(
            String plainEnglish,
            String likelyCause,
            String confidence,
            JsonArray fixHints,
            boolean manualReview) {
    }

    private CrashNarrator() {
    }

    public static Narrative narrate(JsonObject crash, JsonArray mods) {
        return narrate(crash, mods, CrashClassifier.ClassifyContext.empty());
    }

    public static Narrative narrate(JsonObject crash, JsonArray mods, CrashClassifier.ClassifyContext ctx) {
        String exception = str(crash, "exception");
        String root = str(crash, "root_exception");
        String causedBy = str(crash, "caused_by");
        String modFile = str(crash, "mod_file");
        String summary = str(crash, "summary");
        String failure = str(crash, "failure_message");
        String description = str(crash, "description");
        String file = str(crash, "file");
        String time = str(crash, "time");
        Integer watchdogMs = crash.has("watchdog_tick_ms") && !crash.get("watchdog_tick_ms").isJsonNull()
                ? crash.get("watchdog_tick_ms").getAsInt() : null;

        String combined = ((exception != null ? exception : "") + " "
                + (modFile != null ? modFile : "") + " "
                + (summary != null ? summary : "") + " "
                + (failure != null ? failure : "") + " "
                + (description != null ? description : "")).toLowerCase(Locale.ROOT);

        CrashClassifier.ClassifyContext classifyCtx = ctx;
        if (classifyCtx == null || classifyCtx.mixinIndex() == null) {
            MixinConfigIndex index = MixinConfigIndex.fromMods(mods);
            classifyCtx = new CrashClassifier.ClassifyContext(
                    mods,
                    index,
                    classifyCtx != null && classifyCtx.bootFailed(),
                    classifyCtx != null ? classifyCtx.classIndex() : null);
        }
        CrashClassifier.Classification classification = CrashClassifier.classify(crash, classifyCtx);
        String suspect = classification.suspectModId();
        if (suspect == null) {
            suspect = classification.primaryModId();
        }
        if (suspect == null && modFile != null && !modFile.isBlank()) {
            suspect = CrashClassifier.sanitizeModId(modFile);
        }
        String failureKind = classification.failureKind();
        String rowFailureKind = str(crash, "failure_kind");
        if (CrashClassifier.FK_WATCHDOG_FOLLOWUP.equals(rowFailureKind)) {
            failureKind = rowFailureKind;
        }
        String stallMod = classification.stallModId();

        if (CrashClassifier.FK_MOD_LOAD_MIXIN.equals(failureKind)) {
            String config = classification.details() != null && classification.details().has("mixin_config")
                    ? classification.details().get("mixin_config").getAsString() : null;
            String plain;
            if (suspect != null && config != null) {
                plain = "Mixin config " + config + " failed while loading mod " + suspect
                        + " — update or remove that mod.";
            } else if (config != null) {
                plain = "Mixin config " + config + " failed to load — identify the owning mod and update or remove it.";
            } else {
                plain = "A mixin config failed to initialise during mod loading.";
            }
            return new Narrative(
                    plain,
                    "Mixin load failure",
                    suspect != null ? "high" : "medium",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_MIXIN_CONFLICT.equals(failureKind)) {
            String conflict = classification.details() != null
                    && classification.details().has("conflict_mod_id")
                    ? classification.details().get("conflict_mod_id").getAsString() : null;
            String plain = conflict != null && suspect != null
                    ? "Mixin conflict between " + suspect + " and " + conflict + " — update both mods."
                    : "Two mods' mixins conflict during apply — update or remove one of them.";
            return new Narrative(plain, "Mixin conflict", "high", classification.fixHints(), false);
        }

        if (CrashClassifier.FK_MOD_LOAD_CONFIG.equals(failureKind)) {
            return new Narrative(
                    "A mod SERVER config (.toml) failed to parse — delete or repair the corrupt config file.",
                    "Corrupt server config",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_ASSET.equals(failureKind)) {
            String loc = classification.details() != null && classification.details().has("invalid_location")
                    ? classification.details().get("invalid_location").getAsString() : "a resource path";
            return new Narrative(
                    "Invalid resource location " + loc + " — fix illegal characters in the datapack or mod asset.",
                    "Invalid asset path",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_DUPLICATE.equals(failureKind)) {
            return new Narrative(
                    "Duplicate mod jars blocked boot — remove the duplicate from mods/.",
                    "Duplicate mods",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_WORLDGEN.equals(failureKind)) {
            return new Narrative(
                    "Worldgen feature order cycle — remove or update conflicting biome/terrain mods.",
                    "Worldgen feature cycle",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_COMPAT.equals(failureKind)) {
            return new Narrative(
                    "FerriteCore neighbor-table access failed — set populateNeighborTable to false temporarily.",
                    "FerriteCore compat",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_ECOSYSTEM.equals(failureKind)) {
            String mod = suspect != null ? suspect : "Create ecosystem";
            return new Narrative(
                    "Ecosystem mismatch for " + mod + " — align Create/Railways/addon versions or update the gated mod.",
                    "Mod ecosystem mismatch",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_SCRIPT.equals(failureKind)) {
            return new Narrative(
                    "KubeJS datapack failed to parse — fix or remove the broken script under kubejs/data.",
                    "KubeJS script error",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_MOD_LOAD_DEPENDENCY.equals(failureKind)) {
            String modLabel = suspect != null ? suspect : "a mod";
            return new Narrative(
                    "NeoForge failed while loading " + modLabel
                            + " — often a version mismatch or missing language provider / dependency.",
                    "Mod failed to load",
                    suspect != null ? "high" : "medium",
                    classification.fixHints().size() > 0
                            ? classification.fixHints()
                            : hintsModLoad(suspect, failure),
                    false);
        }

        if (CrashClassifier.FK_PLATFORM_MISMATCH.equals(failureKind)) {
            return new Narrative(
                    "A mod requires a newer Java version than this server JVM — upgrade Java or use an older mod build.",
                    "Java version mismatch",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_ENV_LOCK.equals(failureKind)) {
            return new Narrative(
                    "A file is locked by another Windows process — close other Minecraft/Java instances and retry.",
                    "File in use",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (CrashClassifier.FK_WORLD_NBT_CORRUPT.equals(failureKind)) {
            return new Narrative(
                    "World or chunk NBT data looks corrupt (ZLIB/EOF while loading). Restore the affected region from a backup.",
                    "Corrupt world data",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (isWatchdog(combined, exception, root)
                || CrashClassifier.FK_WATCHDOG.equals(failureKind)
                || CrashClassifier.FK_WATCHDOG_PREGEN.equals(failureKind)
                || CrashClassifier.FK_WATCHDOG_FOLLOWUP.equals(failureKind)) {
            int ms = watchdogMs != null ? watchdogMs : 60000;
            int sec = Math.max(1, ms / 1000);
            if (CrashClassifier.FK_WATCHDOG_PREGEN.equals(failureKind)) {
                String stall = stallMod != null ? stallMod : "map render";
                return new Narrative(
                    "Server tick hang — " + stall + " blocked while Chunky pregen was active (~" + sec
                            + "s). Pause pregen or defer map render.",
                    "Tick hang / pregen contention",
                    "high",
                    classification.fixHints(),
                    false);
            }
            return new Narrative(
                    "The main server thread stopped responding for ~" + sec + "s (tick watchdog). "
                            + "Read the thread dump first — lag, pregen, or a heavy assembly, not always a broken mod.",
                    "Server hung",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if (isOom(combined) || (CrashClassifier.FK_HOST_RESOURCE.equals(failureKind)
                && classification.details() != null && classification.details().has("oom_kind"))) {
            String oomKind = classification.details() != null && classification.details().has("oom_kind")
                    ? classification.details().get("oom_kind").getAsString() : "heap";
            if ("native".equals(oomKind)) {
                return new Narrative(
                        "The JVM ran out of native or direct memory.",
                        "Out of native memory",
                        "high",
                        classification.fixHints(),
                        false);
            }
            return new Narrative(
                    "Java ran out of heap memory during play.",
                    "Out of memory",
                    "high",
                    classification.fixHints(),
                    false);
        }

        if ("mod".equals(classification.category()) && (suspect != null || classification.primaryModId() != null)
                && CrashClassifier.FK_MOD_RUNTIME.equals(failureKind)) {
            String mod = suspect != null ? suspect : classification.primaryModId();
            JsonObject details = classification.details();
            String createIssue = details != null && details.has("create_issue") && !details.get("create_issue").isJsonNull()
                    ? details.get("create_issue").getAsString() : null;
            String plain;
            if ("create".equals(mod) && CrashClassifier.CREATE_ISSUE_CONTRAPTION.equals(createIssue)) {
                plain = "Create contraption collision (" + mod
                        + ") — stop the stuck assembly so the world can load, then update Create if needed.";
            } else if ("create".equals(mod)) {
                plain = "Create crashed during play (" + mod
                        + ") — inspect the stack and update Create or matching addons if versions look wrong.";
            } else {
                plain = "The crash points to mod " + mod + " — check for updates, corrupt jars, or mixin conflicts.";
            }
            return new Narrative(
                    plain,
                    "Mod crash",
                    "medium",
                    classification.fixHints(),
                    false);
        }

        if (isModLoad(combined, failure, exception) || ("mod".equals(classification.category())
                && failureKind != null && failureKind.startsWith("mod_load_"))) {
            String modLabel = suspect != null ? suspect : "a mod";
            return new Narrative(
                    "NeoForge failed while loading " + modLabel + " — often a version mismatch or missing dependency.",
                    "Mod failed to load",
                    suspect != null ? "high" : "medium",
                    classification.fixHints().size() > 0
                            ? classification.fixHints()
                            : hintsModLoad(suspect, failure),
                    false);
        }

        if ("host_resource".equals(classification.category())) {
            return new Narrative(
                    "A host or JVM resource limit was hit — review CPU, RAM, and disk around the crash time.",
                    "Host resource limit",
                    "medium",
                    classification.fixHints(),
                    false);
        }

        if ("loader".equals(classification.category())) {
            return new Narrative(
                    "NeoForge or the mod loader failed during bootstrap — often incompatible or corrupt mod jars.",
                    "Loader bootstrap failure",
                    "medium",
                    classification.fixHints(),
                    false);
        }

        String known = firstNonBlank(description, failure, causedBy, exception, summary);
        StringBuilder plain = new StringBuilder("We could not determine a specific cause");
        if (file != null && !file.isBlank()) {
            plain.append(" for crash report ").append(file);
        }
        if (time != null && !time.isBlank()) {
            plain.append(" (").append(time).append(")");
        }
        plain.append(".");
        if (known != null && !known.isBlank()) {
            plain.append(" The report mentions: ").append(truncate(known, 160)).append(".");
        }

        return new Narrative(
                plain.toString(),
                "Unknown",
                "low",
                hintsManualReview(),
                true);
    }

    public static void enrichSummary(JsonObject row, Narrative narrative) {
        row.addProperty("plain_english", narrative.plainEnglish());
        row.addProperty("likely_cause", narrative.likelyCause());
        row.addProperty("confidence", narrative.confidence());
        row.addProperty("manual_review", narrative.manualReview());
        row.add("fix_hints", narrative.fixHints());
    }

    private static boolean isWatchdog(String combined, String exception, String root) {
        return combined.contains("serverhangwatchdog")
                || (exception != null && exception.contains("ServerHangWatchdog"))
                || (root != null && root.contains("ServerHangWatchdog"));
    }

    private static boolean isOom(String combined) {
        return combined.contains("outofmemoryerror") || combined.contains("java heap space")
                || combined.contains("gc overhead limit");
    }

    private static boolean isModLoad(String combined, String failure, String exception) {
        return combined.contains("mod loading has failed")
                || combined.contains("modloadingcrash")
                || combined.contains("modloadingexception")
                || (exception != null && exception.contains("ModLoading"));
    }

    private static JsonArray hintsWatchdog() {
        return toArray(List.of(
                "Pause Chunky or Distant Horizons pregen if it was running",
                "Open Spark or a tick profiler; reduce simulation distance if MSPT was high",
                "If this repeats with no players online, find chunk loaders or rogue entities"));
    }

    private static JsonArray hintsOom() {
        return toArray(List.of(
                "Increase Java heap (-Xmx) if headroom is low",
                "Remove duplicate mods or oversized pregen jobs that leak memory",
                "Run Spark heap analysis if OOM repeats under normal load"));
    }

    private static JsonArray hintsModLoad(String suspect, String failure) {
        List<String> hints = new ArrayList<>();
        if (suspect != null) {
            hints.add("Update or reinstall " + suspect + " from Modrinth or the official source");
            hints.add("Check latest.log for missing dependencies for " + suspect);
        } else {
            hints.add("Open latest.log and find which mod failed to load");
        }
        if (failure != null && failure.toLowerCase(Locale.ROOT).contains("dependency")) {
            hints.add("Install or update the dependency mod cited in the failure message");
        }
        hints.add("Remove recently added mods one at a time until the server starts");
        return toArray(hints);
    }

    private static JsonArray hintsManualReview() {
        return toArray(List.of(
                "Open the full crash report under crash-reports/ and read the root exception",
                "Search the mod id or exception online or in your pack issue tracker",
                "Mark reviewed after you confirm the crash is resolved or historical"));
    }

    private static JsonArray toArray(List<String> hints) {
        JsonArray arr = new JsonArray();
        hints.forEach(arr::add);
        return arr;
    }

    private static String firstNonBlank(String... values) {
        for (String v : values) {
            if (v != null && !v.isBlank()) {
                return v;
            }
        }
        return null;
    }

    private static String truncate(String s, int max) {
        return s.length() > max ? s.substring(0, max) + "…" : s;
    }

    private static String str(JsonObject o, String key) {
        if (o == null || !o.has(key) || o.get(key).isJsonNull()) {
            return null;
        }
        return o.get(key).getAsString();
    }
}

File: web/dr-viewer/analyze/crashNarrator.js
/**
 * Port of CrashNarrator.java (1.0.13 + 1.0.16 CA parity failure_kind narratives).
 */
import {
  classifyCrash,
  FK_WORLD_NBT_CORRUPT,
  FK_WATCHDOG,
  FK_WATCHDOG_PREGEN,
  FK_WATCHDOG_FOLLOWUP,
  FK_MOD_RUNTIME,
  FK_HOST_RESOURCE,
  FK_MOD_LOAD_MIXIN,
  FK_MOD_LOAD_MIXIN_CONFLICT,
  FK_MOD_LOAD_DUPLICATE,
  FK_MOD_LOAD_CONFIG,
  FK_MOD_LOAD_ASSET,
  FK_MOD_LOAD_WORLDGEN,
  FK_MOD_LOAD_COMPAT,
  FK_MOD_LOAD_ECOSYSTEM,
  FK_MOD_LOAD_SCRIPT,
  FK_MOD_LOAD_DEPENDENCY,
  FK_PLATFORM_MISMATCH,
  FK_ENV_LOCK,
  sanitizeModId,
} from './crashClassifier.js';

function str(o, key) {
  return o?.[key] ?? null;
}

function isWatchdog(combined, exception, root) {
  return combined.includes('serverhangwatchdog')
    || exception?.includes('ServerHangWatchdog')
    || root?.includes('ServerHangWatchdog');
}

function isOom(combined) {
  return combined.includes('outofmemoryerror')
    || combined.includes('java heap space')
    || combined.includes('gc overhead limit');
}

function isModLoad(combined, failure, exception) {
  return combined.includes('mod loading has failed')
    || combined.includes('modloadingcrash')
    || combined.includes('modloadingexception')
    || (failure && failure.trim())
    || exception?.includes('ModLoading');
}

function hintsWatchdog() {
  return [
    'Read the watchdog thread dump — the stuck stack names the hang (mod, worldgen, or farm).',
    'Pause Chunky / Distant Horizons / map render only if pregen or those mods appear in the dump.',
    'If MSPT was high, reduce simulation distance or find chunk loaders / rogue entities.',
  ];
}

function hintsOom() {
  return [
    'Confirm the pack needs more heap before raising RAM — oversized packs and leaks look the same.',
    'Increase Java heap (-Xmx) only if the host still has free RAM; otherwise find leaks or shrink the pack.',
    'Check duplicate mods, oversized chunk loaders, or run Spark heap analysis.',
  ];
}

function hintsModLoad(suspect, failure) {
  const hints = [];
  if (suspect) {
    hints.push(`Update or reinstall ${suspect} from Modrinth or the official source`);
    hints.push(`Check latest.log for missing dependencies for ${suspect}`);
  } else {
    hints.push('Open latest.log and find which mod failed to load');
  }
  if (failure?.toLowerCase().includes('dependency')) {
    hints.push('Install or update the dependency mod cited in the failure message');
  }
  hints.push('Remove recently added mods one at a time until the server starts');
  return hints;
}

function hintsManualReview() {
  return [
    'Open the full crash report under crash-reports/ and read the root exception',
    'Search the mod id or exception online or in your pack issue tracker',
    'Mark reviewed after you confirm the crash is resolved or historical',
  ];
}

function firstNonBlank(...values) {
  for (const v of values) {
    if (v != null && String(v).trim()) return v;
  }
  return null;
}

function truncate(s, max) {
  return s.length > max ? `${s.slice(0, max)}…` : s;
}

export function narrateCrash(crash) {
  const exception = str(crash, 'exception');
  const root = str(crash, 'root_exception');
  const causedBy = str(crash, 'caused_by');
  const modFile = str(crash, 'mod_file');
  const summary = str(crash, 'summary');
  const failure = str(crash, 'failure_message');
  const description = str(crash, 'description');
  const file = str(crash, 'file');
  const time = str(crash, 'time');
  const watchdogMs = crash.watchdog_tick_ms ?? null;

  const combined = `${exception || ''} ${modFile || ''} ${summary || ''} ${failure || ''} ${description || ''}`.toLowerCase();
  const classification = classifyCrash(crash);
  let suspect = classification.suspect_mod_id;
  if (!suspect) suspect = classification.primary_mod_id;
  if (!suspect && modFile) {
    suspect = sanitizeModId(modFile) || modFile.replace(/\.jar$/i, '').split('-')[0].toLowerCase();
  }
  const failureKind = classification.failure_kind;
  const stallMod = classification.stall_mod_id;

  if (failureKind === FK_WORLD_NBT_CORRUPT) {
    return {
      plain_english: 'World or chunk NBT data looks corrupt (ZLIB/EOF while loading). Restore the affected region from a backup.',
      likely_cause: 'Corrupt world data',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_MIXIN) {
    const mod = suspect || 'a mod';
    const cfg = classification.details?.mixin_config;
    return {
      plain_english: cfg
        ? `Mixin config ${cfg} failed to initialise${suspect ? ` (owned by ${mod})` : ''}. Update or temporarily remove the mod.`
        : `A mixin failed to initialise for ${mod}. Update recent mods or check mixin conflicts.`,
      likely_cause: 'Mixin init failure',
      confidence: suspect ? 'high' : 'medium',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_MIXIN_CONFLICT) {
    const a = classification.details?.mixin_config;
    const b = classification.details?.mixin_config_conflict;
    return {
      plain_english: a && b
        ? `Two mixin configs conflict (${a} vs ${b}). Update both mods or remove one.`
        : 'Two mods\' mixins conflict during apply. Update both or remove one.',
      likely_cause: 'Mixin conflict',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_DUPLICATE) {
    return {
      plain_english: 'Duplicate mod jars were found (same mod id twice). Keep only one jar per mod id, then restart.',
      likely_cause: 'Duplicate mods',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_CONFIG) {
    const cfg = classification.details?.config_file || 'SERVER config';
    return {
      plain_english: `Corrupt SERVER config ${cfg}${suspect ? ` for ${suspect}` : ''}. Delete or fix it, then let the mod regenerate defaults.`,
      likely_cause: 'Corrupt mod config',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_ASSET) {
    const loc = classification.details?.invalid_location || 'resource path';
    return {
      plain_english: `Illegal character in resource location ${loc}. Fix the datapack/mod asset name (a-z 0-9 / . _ - only).`,
      likely_cause: 'Invalid asset path',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_WORLDGEN) {
    return {
      plain_english: 'Worldgen feature order cycle — remove or update conflicting biome/terrain mods.',
      likely_cause: 'Worldgen feature cycle',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_COMPAT) {
    return {
      plain_english: 'FerriteCore neighbor-table access conflict. Set populateNeighborTable to false as a temporary workaround.',
      likely_cause: 'FerriteCore compat',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_ECOSYSTEM) {
    const mod = suspect || classification.primary_mod_id || 'ecosystem mod';
    return {
      plain_english: `Mod ecosystem mismatch around ${mod} — align versions of the core mod and its addons.`,
      likely_cause: 'Ecosystem version mismatch',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_PLATFORM_MISMATCH) {
    const jm = classification.details?.java_mismatch;
    const plain = jm
      ? `A mod needs Java ${jm.compiled_java} but the server runs Java ${jm.runtime_java}. Upgrade the JVM or use an older mod build.`
      : 'A mod was compiled for a newer Java than the server JVM. Upgrade Java or use a matching mod build.';
    return {
      plain_english: plain,
      likely_cause: 'Java version mismatch',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_ENV_LOCK) {
    const path = classification.details?.locked_path || 'a file';
    return {
      plain_english: `Windows has ${path} locked by another process. Close other Java/Minecraft instances or antivirus scans, then retry.`,
      likely_cause: 'File lock',
      confidence: 'high',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (failureKind === FK_MOD_LOAD_SCRIPT || failureKind === FK_MOD_LOAD_DEPENDENCY) {
    const modLabel = suspect || 'a mod';
    return {
      plain_english: failureKind === FK_MOD_LOAD_SCRIPT
        ? `Script/datapack load failed for ${modLabel}. Fix or remove the broken script cited in the log.`
        : `NeoForge failed while loading ${modLabel} — often a version mismatch or missing language provider / dependency.`,
      likely_cause: failureKind === FK_MOD_LOAD_SCRIPT ? 'Script load failure' : 'Mod failed to load',
      confidence: suspect ? 'high' : 'medium',
      fix_hints: classification.fix_hints?.length ? classification.fix_hints : hintsModLoad(suspect, failure),
      manual_review: false,
    };
  }

  if (isWatchdog(combined, exception, root)
    || failureKind === FK_WATCHDOG
    || failureKind === FK_WATCHDOG_PREGEN
    || failureKind === FK_WATCHDOG_FOLLOWUP) {
    const ms = watchdogMs ?? 60000;
    const sec = Math.max(1, Math.floor(ms / 1000));
    if (failureKind === FK_WATCHDOG_PREGEN || stallMod) {
      const stall = stallMod || 'map render';
      const hints = classification.fix_hints?.length ? classification.fix_hints : hintsWatchdog();
      return {
        plain_english: `Server tick hang — ${stall} blocked while Chunky pregen was active (~${sec}s). Pause pregen or defer map render.`,
        likely_cause: 'Tick hang / pregen contention',
        confidence: 'high',
        fix_hints: hints,
        manual_review: false,
      };
    }
    return {
      plain_english: `The main server thread stopped responding for ~${sec}s (tick watchdog). Read the thread dump first — lag, pregen, or a heavy assembly, not always a broken mod.`,
      likely_cause: 'Server hung',
      confidence: 'high',
      fix_hints: classification.fix_hints?.length ? classification.fix_hints : hintsWatchdog(),
      manual_review: false,
    };
  }

  if (isOom(combined) || (failureKind === FK_HOST_RESOURCE && classification.details?.oom_kind)) {
    const oomKind = classification.details?.oom_kind || 'heap';
    return {
      plain_english: oomKind === 'native'
        ? 'The JVM ran out of native or direct memory.'
        : 'Java ran out of heap memory during play.',
      likely_cause: oomKind === 'native' ? 'Out of native memory' : 'Out of memory',
      confidence: 'high',
      fix_hints: classification.fix_hints?.length ? classification.fix_hints : hintsOom(),
      manual_review: false,
    };
  }

  if (classification.category === 'mod' && failureKind === FK_MOD_RUNTIME && (suspect || classification.primary_mod_id)) {
    const mod = suspect || classification.primary_mod_id;
    const createIssue = classification.details?.create_issue;
    let plain;
    if (mod === 'create' && createIssue === 'contraption_collision') {
      plain = `Create contraption collision (${mod}) — stop the stuck assembly so the world can load, then update Create if needed.`;
    } else if (mod === 'create') {
      plain = `Create crashed during play (${mod}) — inspect the stack and update Create or matching addons if versions look wrong.`;
    } else {
      plain = `The crash points to mod ${mod} — update it, replace a corrupt jar, or check mixin conflicts.`;
    }
    return {
      plain_english: plain,
      likely_cause: 'Mod crash',
      confidence: 'medium',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (isModLoad(combined, failure, exception)) {
    const modLabel = suspect || 'a mod';
    return {
      plain_english: `NeoForge failed while loading ${modLabel} — often a version mismatch or missing dependency.`,
      likely_cause: 'Mod failed to load',
      confidence: suspect ? 'high' : 'medium',
      fix_hints: classification.fix_hints?.length ? classification.fix_hints : hintsModLoad(suspect, failure),
      manual_review: false,
    };
  }

  if (classification.category === 'host_resource') {
    return {
      plain_english: 'A host or JVM resource limit was hit — review CPU, RAM, and disk around the crash time.',
      likely_cause: 'Host resource limit',
      confidence: 'medium',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  if (classification.category === 'loader') {
    return {
      plain_english: 'NeoForge or the mod loader failed during bootstrap — often incompatible or corrupt mod jars.',
      likely_cause: 'Loader bootstrap failure',
      confidence: 'medium',
      fix_hints: classification.fix_hints,
      manual_review: false,
    };
  }

  const known = firstNonBlank(description, failure, causedBy, exception, summary);
  let plain = 'We could not determine a specific cause';
  if (file) plain += ` for crash report ${file}`;
  if (time) plain += ` (${time})`;
  plain += '.';
  if (known) plain += ` The report mentions: ${truncate(String(known), 160)}.`;

  return {
    plain_english: plain,
    likely_cause: 'Unknown',
    confidence: 'low',
    fix_hints: hintsManualReview(),
    manual_review: true,
  };
}

export function enrichSummary(row, narrative) {
  row.plain_english = narrative.plain_english;
  row.likely_cause = narrative.likely_cause;
  row.confidence = narrative.confidence;
  row.manual_review = narrative.manual_review;
  row.fix_hints = narrative.fix_hints;
}
